cmd/create_test.go

```diff
@@ -100,78 +82,26 @@ func TestCreate_ValidatesName(t *testing.T) {
 	}
 }
 
-// TestCreate_RepositoriesPath ensures that the create command utilizes the
-// expected repositories path, respecting the setting for XDG_CONFIG_PATH
+// TestCreateConfig_RepositoriesPath ensures that the create command utilizes
+// the expected repositories path, respecting the setting for XDG_CONFIG_PATH
 // when deriving the default
-func TestCreate_RepositoriesPath(t *testing.T) {
-	defer fromTempDir(t)()
+func TestCreateConfig_RepositoriesPath(t *testing.T) {
+	defer Fromtemp(t)()
 
 	// Update XDG_CONFIG_HOME to point to some arbitrary location.
-	xdgConfigHome, err := ioutil.TempDir("", "alice")
-	if err != nil {
-		t.Fatal(err)
-	}
-	os.Setenv("XDG_CONFIG_HOME", xdgConfigHome)
+	xdgConfigHome := t.TempDir()
+	defer WithEnvVar(t, "XDG_CONFIG_HOME", xdgConfigHome)()
 
-	// The expected full path to repositories:
+	// The expected full path is XDG_CONFIG_HOME/func/repositories
 	expected := filepath.Join(xdgConfigHome, "func", "repositories")
 
-	// Create command takes a function which will be invoked with the final
-	// state of the createConfig, usually used to do fn.Client instantiation
-	// after flags, environment variables, etc. are calculated.  In this case it
-	// will validate the test condition:  that config reflects the value of
-	// XDG_CONFIG_HOME, and secondarily the path suffix `func/repositories`.
-	cmd := NewCreateCmd(func(cfg ClientOptions) *fn.Client {
-		if cfg.RepositoriesPath != expected {
```

Comment:
Inability to do this test is exactly why I don't like using `fn.Options` which is totally opaque.
But I guess I can live with it.

Reply:
I understand the frustration, and indeed sympathize with it!  However this is actually as indended:  XDG_CONFIG_HOME is an implementation detail of the `func` (core) package, and is therefore opaque to the command package.  Therefore tests can only verify it by side-effect to avoid coupling `cmd` and `func` packages.  This is the same reason why tests in the `func` package are in actually in the `func_test` package.

It would actually be better to refactor the system such that environment variables are never used from anywhere outside of `main`, with their values plumbed through as needed.  This would both improve the determinism of the system as well as make it the responsibility of the `cmd` (`main`) package to test.

Reply:
>XDG_CONFIG_HOME is an implementation detail of the func (core) package, and is therefore opaque to the command package.

But then handing of `XDG_CONFIG_HOME` should be tested (and probably is) in core package and not here.

IMHO `cmd` pacakge can be free of dependency on `fn` core pacakge.
At least test should use only mock not actual fn.Client.